fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -338,7 +342,20 @@ void FixedLagSmoother::processQueue(fuse_core::Transaction& transaction)
   while (transaction_riter != pending_transactions_.rend())
   {
     auto& element = *transaction_riter;
-    if (std::find(sensor_blacklist.begin(), sensor_blacklist.end(), element.sensor_name) != sensor_blacklist.end())
+    auto min_stamp = element.stamp();
+    if (!element.transaction->involvedStamps().empty() && *element.transaction->involvedStamps().begin() < min_stamp)
```

Comment:
I wonder what's the relation between the transaction `stamp()` and the `involvedStamps()`. IMHO it's probably better to make the transaction `stamp()` return `involved_stamps_.begin()` or `ros::Time(0, 0)` if `involved_stamps_.empty()`. I'm not sure if there's a valid case where `stamp()` could or should be different than any those two cases. :thinking: 

Why not only checking the involved stamps? Would it matter if there's no involved stamp? I believe that's already handled in: https://github.com/locusrobotics/fuse/blob/3b59771f2a71e3e67c6cc5f063b7d7c6ef716ad0/fuse_core/src/timestamp_manager.cpp#L66-L70

Reply:
Maybe add this method to `Transaction`:

```c++
const ros::Time& Transaction::oldestStamp() const
{
  if (involved_stamps_.empty())
  {
    return stamp_;
  }
  else
  {
    return std::min(*involved_stamps_.begin(), stamp_);
  }
}
```

and this to the `TransactionQueueElement`:
``` c++
const ros::Time& oldestStamp() const { return transaction->oldestStamp(); }
```

So you can do this here: `const auto& min_stamp = element.oldestStamp();`

If we get rid of `stamp_` though, that could be further simplified, but I'm still not sure about that part. The question is, should the transaction stamp always be the oldest involved stamps? Or zero if there's no involved stamp, which is the current behaviour. :thinking:

Reply:
I like the concept of the transaction having an "oldestStamp()" method. That cleans up this code nicely, and insulates it from future changes.

As far as the Transaction stamp vs involved_stamps. I'm honestly not sure what the correct answer is. In my mind, I want the transactions applied in the order that they were measured/computed/etc. So in the case of relative poses, the pose1->pose2 measurement will occur at the pose2 timestamp. So the transaction stamp == pose2 stamp, but the involved stamps will contain {pose1 stamp, pose2 stamp}. But I'm honestly not sure if the application order matters at all, it just felt like the thing to do at the time.

I guess one small difference will be transaction timeout. If we always use the oldest stamp, a transaction could timeout instantly. We could add a `newestStamp()` method to the transaction as well, and use the newest stamp for timeout purposes.

Reply:
> I guess one small difference will be transaction timeout. If we always use the oldest stamp, a transaction could timeout instantly. We could add a newestStamp() method to the transaction as well, and use the newest stamp for timeout purposes.

Yes, I added `newestStamp()` in #176 , but the transaction timeout logic I implemented is actually comparing against the oldest stamp of each transaction:
- https://github.com/locusrobotics/fuse/pull/176/files#diff-db29a07a8b4cc5e7dfa08f70ca3907aaR337
- https://github.com/locusrobotics/fuse/pull/176/files#diff-db29a07a8b4cc5e7dfa08f70ca3907aaR372

I think you're right. The timeout logic should use the `newestStamp()` of all transactions. That would required more changes, since we can't rely on the `pending_transactions_` ordering, if it's based on the `oldestStamp()`, as I did in #176

Reply:
This should be good for now, for this PR. I think this can be merged as is and we can continue the discussion and fixed the remaining issues in #176.